[dtensor][MTPG] make sharding prop lru cache not shared among threads

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134294


**Summary**
Before this PR, `sharding propagator` is shared among threads. The result is the cache result of rank 0 would be accessible by other ranks e.g. rank 1 and this could lead to wrong DTensor resharding. This PR fixes it by making the cache a local variable at thread level, and it fixes `dstack` test (#126493), `inner` (https://github.com/pytorch/pytorch/issues/126852), and `vstack` (https://github.com/pytorch/pytorch/issues/126868). It also fixes `poisson_nll` (https://github.com/pytorch/pytorch/issues/131446) as a bi-product. 

**Test**
`pytest test/distributed/_tensor/test_dtensor_ops.py`


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

XilunWu committed:
[dtensor][MTPG] make sharding prop lru cache not shared among threads

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][MTPG] make sharding prop lru cache not shared among threads"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

tianyu-l commented on torch/distributed/_tensor/_sharding_prop.py:
```diff
@@ -37,6 +38,14 @@ def _length(obj) -> int:
+    def __init__(self, user_function: Callable) -> None:
+        self.cache = lru_cache(None)(user_function)
+
+    def apply(self, *args, **kwargs) -> object:
```
can we instead overwrite the `__call__` method, so that we can save this `.apply` and still use `self.propagate_op_sharding(op_info.schema)`?

XilunWu replied:
yeah that's better.

wz337 reviewed (approved):
Nice. MTPG issues like this are annoying to debug.

wz337 commented:
@XilunWu Could you do me a favor to close the corresponding issues and mark it completed on this main issue page when you land? https://github.com/pytorch/pytorch/issues/132845

XilunWu committed:
Update on "[dtensor][MTPG] make sharding prop lru cache not shared among threads"



**Summary**
Before this PR, `sharding propagator` is shared among threads. The result is the cache result of rank 0 would be accessible by other ranks e.g. rank 1 and this could lead to wrong DTensor resharding. This PR fixes it by making the cache a local variable at thread level, and it fixes `dstack` test (#126493), `inner` (https://github.com/pytorch/pytorch/issues/126852), and `vstack` (https://github.com/pytorch/pytorch/issues/126868). It also fixes `poisson_nll` (https://github.com/pytorch/pytorch/issues/131446) as a bi-product. 

**Test**
`pytest test/distributed/_tensor/test_dtensor_ops.py`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

awgu commented:
Great work!

Failure looks real -- could you fix this before landing:
```
2024-08-23T10:39:54.1440077Z Traceback (most recent call last):
2024-08-23T10:39:54.1440787Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 664, in run_test
2024-08-23T10:39:54.1441425Z     getattr(self, test_name)()
2024-08-23T10:39:54.1442100Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 543, in wrapper
2024-08-23T10:39:54.1442716Z     fn()
2024-08-23T10:39:54.1443311Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2918, in wrapper
2024-08-23T10:39:54.1443915Z     method(*args, **kwargs)
2024-08-23T10:39:54.1444650Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 378, in wrapper
2024-08-23T10:39:54.1445377Z     raise e
2024-08-23T10:39:54.1446062Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 375, in wrapper
2024-08-23T10:39:54.1446853Z     func(self, *args, **kwargs)  # type: ignore[misc]
2024-08-23T10:39:54.1447657Z   File "/var/lib/jenkins/pytorch/test/distributed/_tensor/test_tensor_ops.py", line 620, in test_dtensor_dtype_conversion
2024-08-23T10:39:54.1448445Z     hits, misses, _, _ = _get_sharding_prop_cache_info()
2024-08-23T10:39:54.1449437Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/distributed/_tensor/debug/__init__.py", line 18, in _get_sharding_prop_cache_info
2024-08-23T10:39:54.1450456Z     DTensor._op_dispatcher.sharding_propagator.propagate_op_sharding.cache_info()  # type:ignore[attr-defined]
2024-08-23T10:39:54.1451165Z AttributeError: 'LocalLRUCache' object has no attribute 'cache_info'
```

XilunWu committed:
Update on "[dtensor][MTPG] make sharding prop lru cache not shared among threads"



**Summary**
Before this PR, `sharding propagator` is shared among threads. The result is the cache result of rank 0 would be accessible by other ranks e.g. rank 1 and this could lead to wrong DTensor resharding. This PR fixes it by making the cache a local variable at thread level, and it fixes `dstack` test (#126493), `inner` (https://github.com/pytorch/pytorch/issues/126852), and `vstack` (https://github.com/pytorch/pytorch/issues/126868). It also fixes `poisson_nll` (https://github.com/pytorch/pytorch/issues/131446) as a bi-product. 

**Test**
`pytest test/distributed/_tensor/test_dtensor_ops.py`


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "seems to have introduced https://github.com/pytorch/pytorch/actions/runs/10537099590/job/29201744658" -c nosignal

XilunWu commented:
interesting, the failing tests are under `xfail`. Running tests locally is fine. Curious why CI doesn't catch `xfail` signals.

XilunWu commented:
Close this PR as it is re-landed in https://github.com/pytorch/pytorch/pull/134509
